### PR TITLE
Expand MPI functionality to simplify logic in Jayenne

### PR DESCRIPTION
**Pull request by alexrlongne**
* Add more MPI functions to Draco's C4 library and allow custom types in send/receives
  * Add non-blocking sends and receives with custom MPI datatypes
  * Add no-blocking allreduce call with the standard of
  * send_is test now includes example of a custom MPI datatype and a send/receive with that type
  * tstReduction now includes non-blocking allreduce with several types

* Purpose of Pull Request
  * Allow MPI messaging for any class without packing and unpacking into character streams
  * Allow Jayenne to use the simpler non-blocking allreduce instead of complicated reduction

* [Pre-Merge Code Review](https://github.com/lanl/Draco/wiki/Style-Guide)
  * [ ] Travis CI checks pass
  * [ ] Code coverage does not decrease
  * [ ] [Valgrind test passes](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [ ] [Toss2 checks pass](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [ ] [Toss3 checks pass](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [ ] [Trinitite checks pass](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [ ] Code reviewed/approved, sufficient DbC checks, testing, documentation
  * [ ] Ensure `cmake -DDRACO_C4=SCALAR ~/draco` works (It doesn't.  Not even before this PR.  KT will help with fixing this issue.)

### CODE REVIEW

**Comment by alexrlongne** [comment]
@KineticTheory Could you review this? Once it's merged I'll submit the matching Jayenne PR.

File: src/c4/C4_MPI.i.hh [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+//---------------------------------------------------------------------------//
+template <typename T>
+DLL_PUBLIC_c4 int send_custom(const T *buffer, int size, int destination,
+                              int tag) {
+  MPI_Send(const_cast<T *>(buffer), size, T::MPI_Type, destination, tag,
+           communicator);
+  return C4_SUCCESS;
+}
+
```
Comment (KineticTheory) [inline src/c4/C4_MPI.i.hh]: Please add a unit test for `int send_custom`, `int receive_custom` and `int message_size_custom`.

File: src/c4/C4_MPI.i.hh [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+namespace rtt_c4 {
+
+//---------------------------------------------------------------------------//
+template <typename T>
+DLL_PUBLIC_c4 void send_is_custom(C4_Req &request, const T *buffer, int size,
+                                  int destination, int tag) {
+
+  Require(!request.inuse());
+
```
Comment (KineticTheory) [inline src/c4/C4_MPI.i.hh]: What happens if `size=0` or `buffer=nullptr`?  Can DbC be added for those cases? Is special logic needed?  (here and new functions below).

**Comment by KineticTheory** [comment]
Just for the record - As you know, something looks wrong on trinitite (cray-mpich).

**Comment by KineticTheory** [comment]
@along I want to see the trinitite failures fixed before I merge this.  The out-of-disk space error last night prevented these tests from running.  I will run them by hand this morning and then finish the merge.

**Comment by alexrlongne** [comment]
@KineticTheory That's fine! I've freed up some disk space in my directory.

**Comment by KineticTheory** [comment]
@along Trinitite is on DST. Ugh.  I will merge this and we can deal with the bugs (if any) later.

**Comment by alexrlongne** [comment]
@KineticTheory The good news is that the tests failing on ccs machines went away with the last commit so I'm pretty confident it cleared up the same three failures on trinitite.

**Pull request merged**